Maui ebook mvvm (#30073)
dotnet/docs: docs/architecture/maui/mvvm.md

---
title: Model-View-ViewModel
description: Overview of the Model-View-ViewModel pattern used by .NET MAUI
author: michaelstonis
no-loc: [MAUI]
ms.date: 07/01/2022
---

> # MVVM

Reviewer: ```suggestion
# Model-View-ViewModel (MVVM)
```

The first reference should be spelled out with the abbreviation.

The .NET MAUI developer experience typically involves creating a user interface in XAML, and then adding code-behind that operates on the user interface. Complex maintenance issues can arise as apps are modified and grow in size and scope. These issues include the tight coupling between the UI controls and the business logic, which increases the cost of making UI modifications, and the difficulty of unit testing such code.

> The Model-View-ViewModel (MVVM) pattern helps cleanly separate an application's business and presentation logic from its user interface (UI). Maintaining a clean separation between application logic and the UI helps address numerous development issues and makes an application easier to test, maintain, and evolve. It can also significantly improve code re-use opportunities and allows developers and UI designers to collaborate more easily when developing their respective parts of an app.

Reviewer: ```suggestion
The MVVM pattern helps cleanly separate an application's business and presentation logic from its user interface (UI). Maintaining a clean separation between application logic and the UI helps address numerous development issues and makes an application easier to test, maintain, and evolve. It can also significantly improve code re-use opportunities and allows developers and UI designers to collaborate more easily when developing their respective parts of an app.
```

- Designers and developers can work independently and concurrently on their components during development. Designers can focus on the view, while developers can work on the view model and model components.

The key to using MVVM effectively lies in understanding how to factor app code into the correct classes and how the classes interact. The following sections discuss the responsibilities of each of the classes in the MVVM pattern.

### View

The view is responsible for defining the structure, layout, and appearance of what the user sees on screen. Ideally, each view is defined in XAML, with a limited code-behind that does not contain business logic. However, in some cases, the code-behind might contain UI logic that implements visual behavior that is difficult to express in XAML, such as animations.

In a .NET MAUI application, a view is typically a Page-derived or ContentView-derived class. However, views can also be represented by a data template, which specifies the UI elements to be used to visually represent an object when it's displayed. A data template as a view does not have any code-behind, and is designed to bind to a specific view model type.

> [!TIP]

> > Avoid enabling and disabling UI elements in the code-behind

Reviewer: ```suggestion
> Avoid enabling and disabling UI elements in the code-behind.
```

All alerts should follow standard sentence casing and punctuation guidelines.

There are several options for executing code on the view model in response to interactions on the view, such as a button click or item selection. If a control supports commands, the control's Command property can be data-bound to an ICommand property on the view model. When the control's command is invoked, the code in the view model will be executed. In addition to commands, behaviors can be attached to an object in the view and can listen for either a command to be invoked or the event to be raised. In response, the behavior can then invoke an ICommand on the view model or a method on the view model.

### ViewModel

The view model implements properties and commands to which the view can data bind to, and notifies the view of any state changes through change notification events. The properties and commands that the view model provides define the functionality to be offered by the UI, but the view determines how that functionality is to be displayed.

> [!TIP]

> > Keep the UI responsive with asynchronous operations

Reviewer: ```suggestion
> Keep the UI responsive with asynchronous operations.
```

Mobile apps should keep the UI thread unblocked to improve the user's perception of performance. Therefore, in the view model, use asynchronous methods for I/O operations and raise events to asynchronously notify views of property changes.

The view model is also responsible for coordinating the view's interactions with any model classes that are required. There's typically a one-to-many relationship between the view model and the model classes. The view model might choose to expose model classes directly to the view so that controls in the view can data bind directly to them. In this case, the model classes will need to be designed to support data binding and change notification events.

Each view model provides data from a model in a form that the view can easily consume. To accomplish this, the view model sometimes performs data conversion. Placing this data conversion in the view model is a good idea because it provides properties that the view can bind to. For example, the view model might combine the values of two properties to make it easier to display by the view.

> [!TIP]

> > Centralize data conversions in a conversion layer

Reviewer: ```suggestion
> Centralize data conversions in a conversion layer.
```

It's also possible to use converters as a separate data conversion layer that sits between the view model and the view. This can be necessary, for example, when data requires special formatting that the view model doesn't provide.

> In order for the view model to participate in two-way data binding with the view, its properties must raise the PropertyChanged event. View models satisfy this requirement by implementing the INotifyPropertyChanged interface, and raising the PropertyChanged event when a property is changed.

Reviewer: ```suggestion
In order for the view model to participate in two-way data binding with the view, its properties must raise the `PropertyChanged` event. View models satisfy this requirement by implementing the `INotifyPropertyChanged` interface, and raising the `PropertyChanged` event when a property is changed.
```

> For collections, the view-friendly ObservableCollection<T> is provided. This collection implements collection changed notification, relieving the developer from having to implement the INotifyCollectionChanged interface on collections.

Reviewer: ```suggestion
For collections, the view-friendly `ObservableCollection<T>` is provided. This collection implements collection changed notification, relieving the developer from having to implement the `INotifyCollectionChanged` interface on collections.
```

With view model first composition, the app is conceptually composed of view models, with a service responsible for locating the view for a view model. View model first composition feels more natural to some developers, since the view creation can be abstracted away, allowing them to focus on the logical non-UI structure of the app. In addition, it allows view models to be created by other view models. However, this approach is often complex, and it can become difficult to understand how the various parts of the app are created and associated.

> [!TIP]

> > Keep view models and views independent

Reviewer: ```suggestion
> Keep view models and views independent.
```

The binding of views to a property in a data source should be the view's principal dependency on its corresponding view model. Specifically, don't reference view types, such as Button and ListView, from view models. By following the principles outlined here, view models can be tested in isolation, therefore reducing the likelihood of software defects by limiting scope.

The following sections discuss the main approaches to connecting view models to views.

## Creating a view model declaratively

The simplest approach is for the view to declaratively instantiate its corresponding view model in XAML. When the view is constructed, the corresponding view model object will also be constructed. This approach is demonstrated in the following code example:

```xaml

> <ContentPage ... xmlns:local="clr-namespace:eShop">

Reviewer: ```suggestion
<ContentPage xmlns:local="clr-namespace:eShop">
```

    <ContentPage.BindingContext>
        <local:LoginViewModel />
    </ContentPage.BindingContext>

>     ...

Reviewer: More often than not, readers will copy code verbatim from wherever they find it. Please try to avoid putting code in place that will not compile. However, if you do need to have something that indicates code was omitted, use a valid (and language-appropriate) comment. For example:

**C#**

```csharp
public sealed record struct Point(double X, double Y)
{
    // Omitted for brevity...
}
```

**XAML**

```xaml
<ContentPage>
    <!-- Omitted for brevity... -->
</ContentPage>
```

Author: That is no problem. I wasn't sure what the standard was there. I will follow this going forward though.

>     ...

Reviewer: ```suggestion
    <!-- Omitted for brevity... -->
```

</ContentPage>
```

> When the ContentPage is created, an instance of the LoginViewModel is automatically constructed and set as the view's BindingContext.

Reviewer: ```suggestion
When the `ContentPage` is created, an instance of the `LoginViewModel` is automatically constructed and set as the view's `BindingContext`.
```

This declarative construction and assignment of the view model by the view has the advantage that it's simple, but has the disadvantage that it requires a default (parameter-less) constructor in the view model.

## Creating a view model programmatically

> A view can have code in the code-behind file, resulting in the view model being assigned to its BindingContext property. This is often accomplished in the view's constructor, as shown in the following code example:

Reviewer: ```suggestion
A view can have code in the code-behind file, resulting in the view-model being assigned to its `BindingContext` property. This is often accomplished in the view's constructor, as shown in the following code example:
```

```csharp
public LoginView()
{
    InitializeComponent();
    BindingContext = new LoginViewModel(navigationService);
}
```

> The programmatic construction and assignment of the view model within the view's code-behind has the advantage that it's simple. However, the main disadvantage of this approach is that the view needs to provide the view model with any required dependencies. Using a dependency injection container can help to maintain loose coupling between the view and view model. For more information, see [Dependency injection](dependency-injection.md).

Reviewer: ```suggestion
The programmatic construction and assignment of the view model within the view's code-behind has the advantage that it's simple. However, the main disadvantage of this approach is that the view needs to provide the view model with any required dependencies. Using a dependency injection container can help to maintain loose coupling between the view and view model. For more information, see [Dependency injection in .NET](../../core/extensions/dependency-injection.md).
```

Author: For this one, I will have a chapter in the book around dependency injection. I haven't kicked up a PR for it yet, so for this one, I believe, is a false positive.

## Updating views in response to changes in the underlying view model or model

> All view model and model classes that are accessible to a view should implement the [INotifyPropertyChanged](/dotnet/api/system.componentmodel.inotifypropertychanged) interface. Implementing this interface in a view model or model class allows the class to provide change notifications to any data-bound controls in the view when the underlying property value changes.

Reviewer: ```suggestion
All view model and model classes that are accessible to a view should implement the [<xref:System.ComponentModel.INotifyPropertyChanged> interface. Implementing this interface in a view model or model class allows the class to provide change notifications to any data-bound controls in the view when the underlying property value changes.
```

Are you using the [Docs Authoring Pack](https://docs.microsoft.com/contribute/how-to-write-docs-auth-pack) VS Code extension? If not, you're missing out on a lot of powerful features that help with writing for docs.microsoft.com. In particular there is a feature to add xref links: https://docs.microsoft.com/contribute/how-to-write-links#xref-cross-reference-links

Author: I am. This was the suggestion that it provided when adding a link. If a link points back to the docs site, should I always use xref? It has been wanting to use links like the above. I will look into this more now that I know!

App's should be architected for the correct use of property change notification, by meeting the following requirements:

> - Always raising a PropertyChanged event if a public property's value changes. Do not assume that raising the PropertyChanged event can be ignored because of knowledge of how XAML binding occurs.

Reviewer: ```suggestion
- Always raising a `PropertyChanged` event if a public property's value changes. Do not assume that raising the `PropertyChanged` event can be ignored because of knowledge of how XAML binding occurs.
```

I started updating all of the references to code elements, by wrapping them in a pair of backticks. This should be done for all code element references, please update the remaining ones.

Author: Can do. I noticed in some docs it had it and in others, it didn't. I prefer this myself, so I have no problem adding this to the upcoming chapters sounds good.

- Never raising more than one PropertyChanged event with the same property name argument within a single synchronous invocation of a public method of a class. For example, given a NumberOfItems property whose backing store is the _numberOfItems field, if a method increments _numberOfItems fifty times during the execution of a loop, it should only raise property change notification on the NumberOfItems property once, after all the work is complete. For asynchronous methods, raise the PropertyChanged event for a given property name in each synchronous segment of an asynchronous continuation chain.

A simple way to provide this functionality would be to create an extension of the BindableObject class. In this example, the ExtendedBindableObject class provides change notifications, which is shown in the following code example:

```csharp
public abstract class ExtendedBindableObject : BindableObject
{
    public void RaisePropertyChanged<T>(Expression<Func<T>> property)
    {
        var name = GetMemberInfo(property).Name;
        OnPropertyChanged(name);
    }

    private MemberInfo GetMemberInfo(Expression expression)
    {

>         ...

Reviewer: ```suggestion
        // Omitted for brevity ...
```

    }
}
```

.NET MAUI's BindableObject class implements the INotifyPropertyChanged interface, and provides an OnPropertyChanged method. The ExtendedBindableObject class provides the RaisePropertyChanged method to invoke property change notification, and in doing so uses the functionality provided by the BindableObject class.

View model classes can then derive from the ExtendedBindableObject class. Therefore, each view model class uses the RaisePropertyChanged method in the ExtendedBindableObject class to provide property change notification. The following code example shows how the eShopOnContainers mobile app invokes property change notification by using a lambda expression:

```csharp
public bool IsLogin
{

>     get
>     {
>         return _isLogin;
>     }

Reviewer: ```suggestion
    get => _isLogin;
```

Author: I will want to update the back-end code sample for this one to match, but that shouldn't take too long

    set
    {
        _isLogin = value;
        RaisePropertyChanged(() => IsLogin);
    }
}
```

> Note that using a lambda expression in this way involves a small performance cost because the lambda expression has to be evaluated for each call. Although the performance cost is small and would not normally impact an app, the costs can accrue when there are many change notifications. However, the benefit of this approach is that it provides compile-time type safety and refactoring support when renaming properties.

Reviewer: ```suggestion
Using a lambda expression in this way involves a small performance cost because the lambda expression has to be evaluated for each call. Although the performance cost is small and would not normally impact an app, the costs can accrue when there are many change notifications. However, the benefit of this approach is that it provides compile-time type safety and refactoring support when renaming properties.
```

The eShopOnContainers mobile app includes an [EventToCommandBehavior](/dotnet/communitytoolkit/maui/behaviors/event-to-command-behavior) class which is provided by the MAUI Community toolkit. EventToCommandBehavior executes a command in response to an event occurring. This class derives from the BaseBehavior<View> class so that the behavior can bind to and execute an ICommand specified by a Command property when the behavior is consumed. The following code example shows the EventToCommandBehavior class:

```csharp
/// <summary>
/// The <see cref="EventToCommandBehavior"/> is a behavior that allows the user to invoke a <see cref="ICommand"/> through an event. It is designed to associate Commands to events exposed by controls that were not designed to support Commands. It allows you to map any arbitrary event on a control to a Command.
/// </summary>
public class EventToCommandBehavior : BaseBehavior<VisualElement>
{

>     ...

Reviewer: ```suggestion
    // Omitted for brevity...
```

        }
    }
}
```

The `OnAttachedTo` and `OnDetachingFrom` methods are used to register and deregister an event handler for the event defined in the `EventName` property. Then, when the event fires, the `OnTriggerHandled` method is invoked, which executes the command.

The advantage of using the `EventToCommandBehavior` to execute a command when an event fires, is that commands can be associated with controls that weren't designed to interact with commands. In addition, this moves event-handling code to view models, where it can be unit tested.

## Invoking behaviors from a view

The EventToCommandBehavior is particularly useful for attaching a command to a control that doesn't support commands. For example, the LoginView uses the `EventToCommandBehavior` to execute the `ValidateCommand` when the user changes the value of their password, as shown in the following code:

```xaml
<Entry
    IsPassword="True"
    Text="{Binding Password.Value, Mode=TwoWay}">
    // Omitted for brevity...
    <Entry.Behaviors>
        <mct:EventToCommandBehavior
            EventName="TextChanged"
            Command="{Binding ValidateCommand}" />
    </Entry.Behaviors>

>     // Omitted for brevity...

Reviewer: ```suggestion
    <!-- Omitted for brevity... -->
```
